ipq40xx: add support for EnGenius EMR5000

Yen-Ting-Shen opened the pull request:
SOC:     IPQ4019 / QCA Dakota
        CPU:     Quad-Core ARMv7 Processor rev 5 (v7l) Cortex-A7
        DRAM:    256 MiB
        FLASH:   NAND 128 MiB
        ETH:     Qualcomm Atheros QCA8072 (2 ports)
        WLAN1:   Qualcomm Atheros QCA4019 2.4GHz 802.11bgn 2:2x2
        WLAN2:   Qualcomm Atheros QCA4019 5GHz(channel 100-165) 802.11a/n/ac 2:2x2
        WLAN3:   Qualcomm Atheros QCA9888 5GHz(channel 36-64) 802.11a/n/ac 2:2x2
        INPUT:   RESET Button
        LEDS:    White, Blue, Red, Orange
        USB:     1x 3.0

        Flash instruction:
        Upload openwrt-ipq40xx-generic-engenius_emr5000-squashfs-nand-factory.bin
        via the firmwary upgrade function of the stock firmware WebUI.

Signed-off-by: Yen-Ting-Shen <frank.shen@senao.com>

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts:
```diff
@@ -0,0 +1,306 @@
+				label = "reset";
+				gpios = <&tlmm 18 GPIO_ACTIVE_LOW>;
+				linux,code = <KEY_RESTART>;
+			};
```
Indent is wrong here.

Yen-Ting-Shen replied:
Fixed.

adrianschmutzler commented:
Consider adding label_mac as discussed in your other PR ...

Yen-Ting-Shen commented:
> Consider adding label_mac as discussed in your other PR ...

Added.

Yen-Ting-Shen committed:
ipq40xx: add support for EnGenius EMR5000

SOC:     IPQ4019 / QCA Dakota
CPU:     Quad-Core ARMv7 Processor rev 5 (v7l) Cortex-A7
DRAM:    256 MiB
FLASH:   NAND 128 MiB
ETH:     Qualcomm Atheros QCA8072 (2 ports)
WLAN1:   Qualcomm Atheros QCA4019 2.4GHz 802.11bgn 2:2x2
WLAN2:   Qualcomm Atheros QCA4019 5GHz(channel 100-165) 802.11a/n/ac 2:2x2
WLAN3:   Qualcomm Atheros QCA9888 5GHz(channel 36-64) 802.11a/n/ac 2:2x2
INPUT:   RESET Button
LEDS:    White, Blue, Red, Orange
USB:     1x 3.0

Flash instruction:
Upload openwrt-ipq40xx-generic-engenius_emr5000-squashfs-nand-factory.bin
via the firmwary upgrade function of the stock firmware WebUI.

Signed-off-by: Yen-Ting-Shen <frank.shen@senao.com>

Yen-Ting-Shen commented:
removed indents in the commit message.
rebased the PR.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts:
```diff
@@ -0,0 +1,306 @@
+};
+
+&wifi1 {
+	status = "ok";
```
Please change all cases of "ok" to "okay".

adrianschmutzler commented on target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh:
```diff
@@ -9,6 +9,10 @@ preinit_set_mac_address() {
 		;;
+	engenius,emr5000)
+		base_mac=$(cat /sys/class/net/eth0/address)
+		ip link set dev eth1 address $(macaddr_add "$base_mac" +3)
```
Why is eth0 getting a valid address automagically and eth1 isn't?

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts:
```diff
@@ -0,0 +1,306 @@
+	leds {
+		compatible = "gpio-leds";
+
+		power: white {
```
`led_power: power {`

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-emr5000.dts:
```diff
@@ -0,0 +1,306 @@
+		};
+
+		blue {
+			label = "emr5000:blue";
```
So, what are those leds for? You should use a three-membered pattern like for the white led ...

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -193,6 +193,22 @@ define Device/engenius_eap1300
+	DEVICE_VENDOR := EnGenius
+	DEVICE_MODEL := EMR5000
+	BOARD_NAME := emr5000
+	DEVICE_DTS := qcom-ipq4019-emr5000
```
please use
SOC := qcom-ipq4019
instead of DEVICE_DTS.

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -193,6 +193,22 @@ define Device/engenius_eap1300
+	DEVICE_DTS := qcom-ipq4019-emr5000
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	DEVICE_PACKAGES := ath10k-firmware-qca9888 uboot-envtools
```
please use the matching ath10k-...-ct package instead.

adrianschmutzler commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -142,6 +150,10 @@ case "$FIRMWARE" in
 		;;
+	engenius,emr5000)
+		ath10kcal_extract "0:ART" 0x5000 0x2f20
+		ath10kcal_patch_mac_crc $(macaddr_add "$(cat /sys/class/net/eth0/address)" 1)
```
Have you ever tested what happens if you drop the ath10kcal_patch_mac_crc lines in this file? Will the correct address be assigned anyway (i.e., is the correct address stored in the calibration data?).

adrianschmutzler commented:
Please update for kernel 5.4. Note that I will add a patch to master soon that will consolidate the DTS files in the target, so maybe wait for that one.
